### PR TITLE
Feature/submission logic

### PR DESCRIPTION
**Pull request by thatbudakguy**
refs #65 
- search results updated dynamically as user interacts with form
- histogram, result counts, pagination, etc. all update dynamically
- adds template for ajax rendering of results
- refactors search form & histogram classes to make them more general & interoperable
- adds a template filter to enable dumping python objs as json for easier consumption via js

### CODE REVIEW

**Comment by rlskoeser** [comment]
Documenting what we discussed: the histogram & reactive form seem like good candidates for unit testing, but not sure we need to do that now. But it might be a good idea to figure out what it looks like for PPA before starting similar javascript work on another project.

**Comment by rlskoeser** [comment]
Finally had a chance to try the code out locally. Generally, seems to be very slick and is working great. The form submits when I add, edit or remove text or change publication years.  Publication year histogram and record counts are updating when the results change. Query string also reflects changes to the form submission, which is great.

- collections and sort results aren't triggering the form to submit on Chrome or Safari, but they do work in Firefox; tried making the checkbox visible in Chrome to see if that made any difference, but it didn't seem to matter.
- relevance sort should be enabled/disabled based on the presence of a keyword search term
- can we revise the query string to only include form fields that have values?

File: ppa/archive/templates/archive/snippets/results_list.html [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{% load static humanize ppa_tags %}
```
Comment (rlskoeser) [inline ppa/archive/templates/archive/snippets/results_list.html]: Perhaps a template comment here to indicate this template can be rendered in the full list or displayed via ajax?

File: ppa/archive/templatetags/ppa_tags.py [hunk 2]
```diff
@@ -76,3 +78,10 @@ def solr_highlight(value, autoescape=True):
         mark_safe(part) if EM_TAG_RE.match(part) else esc(part)
         for part in EM_TAG_RE.split(value)
     ]))
+
+
+@register.filter(name='json')
+def json_dumps(data):
```
Comment (rlskoeser) [inline ppa/archive/templatetags/ppa_tags.py]: Annoying we have to do this, but seems like a reasonable approach.

File: ppa/archive/tests/test_views.py [hunk 3]
```diff
@@ -418,13 +418,37 @@ def test_digitizedwork_listview(self):
         assert not response.context['object_list'].count()
         self.assertContains(response, 'Invalid range')
 
+        # ajax request for search results
+        response = self.client.get(url, HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        assert response.status_code == 200
+        # should render the results list partial and single result partial
+        self.assertTemplateUsed('archive/snippets/results_list.html')
+        self.assertTemplateUsed('archive/snippest/search_result.html')
```
Comment (rlskoeser) [inline ppa/archive/tests/test_views.py]: Also check `assertTemplateNotUsed` on `archive/list_digitizedworks.html`

File: ppa/archive/tests/test_views.py [hunk 4]
```diff
@@ -418,13 +418,37 @@ def test_digitizedwork_listview(self):
...
+        # shouldn't render the search form
+        self.assertNotContains(response, '<form class="ui form">', html=True)
+        # should have all the results
+        assert len(response.context['object_list']) == len(digitized_works)
```
Comment (rlskoeser) [inline ppa/archive/tests/test_views.py]: add basic sanity check for the data that needs to be present for the ajax form logic to update count and histogram

File: ppa/archive/tests/test_views.py [hunk 5]
```diff
@@ -418,13 +418,37 @@ def test_digitizedwork_listview(self):
...
+        response = self.client.get(url,
+            {'query': 'blood AND bone AND alternate'},
+            HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertContains(response, '1 digitized work')
+        self.assertContains(response, wintry.source_id)
+        response = self.client.get(url,
+            {'query': 'blood NOT bone'},
+            HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertContains(response, 'No matching works.')
+        response = self.client.get(url,
```
Comment (rlskoeser) [inline ppa/archive/tests/test_views.py]: Not sure how much we gain from testing more than one search query, since it's the same view logic we've already tested pretty thoroughly -- unless there are any variations that impact the ajax behavior or work differently.

File: ppa/archive/templatetags/ppa_tags.py (hunk 2 above)
Comment (rlskoeser) [reply ppa/archive/templatetags/ppa_tags.py]: Needs a unit test

File: ppa/archive/views.py [hunk 6]
```diff
@@ -32,6 +32,12 @@ class DigitizedWorkListView(ListView):
     # keyword query; assume no search terms unless set
     query = None
 
+    def get_template_names(self):
+        # allow querying just the results via ajax
```
Comment (rlskoeser) [inline ppa/archive/views.py]: I would rephrase this comment to make it more specific, something like: when queried via ajax, return partial html for just the results section without the form

File: ppa/archive/static/archive/js/search.js [hunk 7]
```diff
@@ -1,63 +1,71 @@
...
-        $(event.target).find('input').attr('checked', '')
+    function submitForm(state) {        
+        let url = `?${$.param(state)}` // serialize state using $.param to make querystring
+        window.history.pushState(state, 'PPA Archive Search', url) // update the URL bar
+        let req = fetch(`/archive/${url}`, { // create the submission request
+            headers: { // this header is needed to signal ajax request to django
+                'X-Requested-With': 'XMLHttpRequest'
+            }
+        })
+        req.then(res => res.text()).then(html => { // submit the form and get html back
```
Comment (rlskoeser) [inline ppa/archive/static/archive/js/search.js]: Nice. Appreciate the comments here

File: ppa/archive/static/archive/js/search.js [hunk 8]
```diff
@@ -1,63 +1,71 @@
...
+            $$paginationTop.html($(html).find('.pagination').html()) // update the top pagination
+            dateHistogram.data = JSON.parse($(html).find('pre').html()) // update the histogram
+            $$resultsCount.html($(html).find('.data .results-count').html()) // update the results count
+            $$results.html(html) // update the results
+            document.dispatchEvent(new Event('ZoteroItemUpdated', { // notify Zotero of changed results
```
Comment (rlskoeser) [inline ppa/archive/static/archive/js/search.js]: Saw this and remembered to test the Zotero integration - it is updating properly and getting results for current items after the ajax form submission.

Comment (thatbudakguy) [reply ppa/archive/static/archive/js/search.js]: awesome!!

File: ppa/archive/tests/test_views.py (hunk 3 above)
Comment (thatbudakguy) [reply ppa/archive/tests/test_views.py]: I think I tried that but it's not a valid assertion?? I couldn't find an idiomatic way to represent that...but I'll take another look

Comment (rlskoeser) [reply ppa/archive/tests/test_views.py]: Does this not work? https://docs.djangoproject.com/en/1.11/topics/testing/tools/#django.test.SimpleTestCase.assertTemplateNotUsed

File: ppa/archive/static/archive/js/search.js (hunk 8 above)
Comment (rlskoeser) [reply ppa/archive/static/archive/js/search.js]: @thatbudakguy should we suggest updated documentation for the Zotero docs? Not sure if they have an easy mechanism for that. Here's where I found the old code I shared with you before: https://www.zotero.org/support/dev/exposing_metadata

File: ppa/archive/tests/test_views.py (hunk 3 above)
Comment (thatbudakguy) [reply ppa/archive/tests/test_views.py]: huh, that definitely looks right. I think either I was trying `assertNotTemplateUsed`, or I did try that but Django was saying that the template _was_ used...I'll try it again.

File: ppa/archive/static/archive/js/search.js (hunk 8 above)
Comment (thatbudakguy) [reply ppa/archive/static/archive/js/search.js]: at least one of the methods in their sample has this deprecation message on MDN:
>This feature has been removed from the Web standards. Though some browsers may still support it, it is in the process of being dropped. Avoid using it and update existing code if possible; see the compatibility table at the bottom of this page to guide your decision. Be aware that this feature may cease to work at any time.

so, probably worth letting them know at least? do you know if we could contact someone, or file a PR? don't know where the docs are hosted or if they're public...

**Comment by rlskoeser** [comment]
Collection and sort still not firing a form submission for me in Chrome, but since they're working for @thatbudakguy going to merge this in so we can have more people try it out.
